exporter/datadogexporter/config/warn_envvars.go

```diff
@@ -0,0 +1,139 @@
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+)
+
+// futureDefaultConfig returns the future default config we will use from v0.50 onwards.
+func futureDefaultConfig() *Config {
+	return &Config{
+		ExporterSettings: config.NewExporterSettings(config.NewComponentID("datadog")),
+		TimeoutSettings: exporterhelper.TimeoutSettings{
+			Timeout: 15 * time.Second,
+		},
+		RetrySettings: exporterhelper.NewDefaultRetrySettings(),
+		QueueSettings: exporterhelper.NewDefaultQueueSettings(),
+		Metrics: MetricsConfig{
+			SendMonotonic: true,
+			DeltaTTL:      3600,
+			Quantiles:     true,
+			ExporterConfig: MetricsExporterConfig{
+				ResourceAttributesAsTags:             false,
+				InstrumentationLibraryMetadataAsTags: false,
+			},
+			HistConfig: HistogramConfig{
+				Mode:         "distributions",
+				SendCountSum: false,
+			},
+		},
+		Traces: TracesConfig{
+			SampleRate:      1,
+			IgnoreResources: []string{},
+		},
+		SendMetadata:        true,
+		UseResourceMetadata: true,
+	}
+}
+
+// errUsedEnvVar constructs an error instructing users not to rely on
+// the automatic environment variable detection and set the environment variable
+// explicitly on configuration instead.
+func errUsedEnvVar(settingName, envVarName string) error {
+	return fmt.Errorf(
+		"%q will not default to %q's value starting on v0.50.0. Set %s: ${%s} to remove this warning",
+		settingName,
+		envVarName,
+		settingName,
+		envVarName,
+	)
+}
+
+// tagsDiffer checks if the host tags from each configuration are different.
+func tagsDiffer(cfgTags []string, futureTags []string) bool {
+	if len(cfgTags) != len(futureTags) {
+		return true
+	}
+
+	oldCfgSet := map[string]struct{}{}
+	for _, tag := range cfgTags {
+		oldCfgSet[tag] = struct{}{}
+	}
+
+	for _, tag := range futureTags {
+		if _, ok := oldCfgSet[tag]; !ok {
+			// tag is in old but not new
+			return true
+		}
+	}
+
+	return false
+}
+
+// warnUseOfEnvVars returns warnings related to automatic environment variable detection.
+// Right now, we automatically get the value for e.g. the API key from DD_API_KEY, even if
+// the user is not doing `api.key: ${DD_API_KEY}`. We are going to remove this functionality.
+func warnUseOfEnvVars(configMap *config.Map, cfg *Config) (warnings []error) {
+	// We don't see the raw YAML contents from our exporter so, to compare with the new default,
+	// we unmarshal the config map on the future default config and compare the two structs.
+	// Any differences will be due to the change in default configuration.
+	futureCfg := futureDefaultConfig()
+	err := configMap.UnmarshalExact(futureCfg)
+	if err != nil {
+		// This should never happen, since unmarshaling passed with the original config.
+		// If it happens, bail out.
+		return
+	}
+
+	// We could probably do this with reflection but I don't want to risk
+	// an accidental panic so we check each field manually.
+	if cfg.API.Key != futureCfg.API.Key {
```

Comment:
```suggestion
	if cfg.API.Key != futureCfg.API.Key && cfg.API.Key != '' {
```
Otherwise, I think this would also trigger if the user has `api.key: ${SOMETHING_DIFFERENT_THAN_DD_API_KEY}` because the field would be empty in the current config and non-empty in the "future" config.

Reply:
Mmmm nope, actually this isn't true because it would be replaced both in the old and new configs.

Reply:
Yes, in that case it would have the value from `SOMETHING_DIFFERENT_THAN_DD_API_KEY` in both places